nix/keys.nix

```diff
@@ -136,23 +136,64 @@ in
               touch /run/keys/done
             '')
         }
-      '';
 
-    systemd.services.nixops-keys =
-      { enable = config.deployment.keys != {};
-        description = "Waiting for NixOps Keys";
-        wantedBy = [ "keys.target" ];
-        before = [ "keys.target" ];
-        unitConfig.DefaultDependencies = false; # needed to prevent a cycle
-        serviceConfig.Type = "oneshot";
-        serviceConfig.RemainAfterExit = true;
-        script =
+        ${concatStringsSep "\n" (flip mapAttrsToList config.deployment.keys (name: value:
+          # Make sure each key has correct ownership, since the configured owning
+          # user or group may not have existed when first uploaded.
+          ''
+            [[ -f "/run/keys/${name}" ]] && chown '${value.user}:${value.group}' "/run/keys/${name}"
```

Comment:
Perhaps. I was unsure whether to fail noisily, or to treat this as best-effort chowning of keys that both (1) we know about at this point in activation and (2) exist, having been placed here for us before activation even started.

I opted for the latter, so that the absence of one key doesn't screw up ownership of other keys or fail an activation. Also, if a service depends on a specific key, they can now do so via a systemd service dependency, since each key gets a service which is running as long as the key exists (added below).